1-Draft/RFCNNNN-Dispose-Script-Block.md

```diff
@@ -0,0 +1,189 @@
+---
+RFC: NNNN
+Author: Joel Sallow
+Status: Draft
+Area: Language
+Comments Due: 7/29/2019
+---
+
+# Function / Script Cmdlet Disposal
+
+Functions in PowerShell are highly versatile, but don't always have the capabilities to manage resources correctly.
+For example, in .NET there are many types which implement `IDisposable`, advising that their `Dispose()` method should be used to clean them up when they're no longer needed.
+In addition to these, there are some types which manage connections to remote hosts or databases which commonly include a `Close()` method which is often to be used prior to calling `Dispose()`.
+Currently, PowerShell functions do not have an effective way of using these resources in the correct manner.
+
+This is further complicated by the fact that any terminating error in the pipeline completely bypasses the `end` named block.
+As such, it cannot be relied upon for cleanup / disposal steps.
+
+This includes any time a user includes the pipeline cmdlet `Select-Object` with the `-First` parameter.
+Once the desired number of objects has been reached, the cmdlet throws an internal and deliberately hidden exception to halt the pipeline.
+This is highly effective, but can lead to undesirable circumstances if essential cleanup steps are then skipped as a result.
+
+Compiled cmdlets do not suffer this issue, as they can simply implement `IDisposable` and place the necessary steps in a `Dispose()` method of their own.
+The pipeline processor will correctly handle **cmdlets** that require disposal steps, but currently not **functions**.
+
+## Motivation
+
+As a developer,<br />
+I want to be able to ensure I am not leaving resources or connections open when a user utilizes a function I have written.
+
+As an administrator,<br />
+I want to be able to ensure that any necessary logging or other necessary administrative tasks cannot be skipped accidentally or deliberately by a user invoking a function in a pipeline and using `Select-Object -First X` to bypass logging which would normally take place in the `end` block.
+
+## Specification
+
+A `dispose{}` block is required for these use cases.
+
+### Execution
+
+The `dispose{}` block will execute:
+
+1. After `end{}` during normal command operation, and
+2. During command disposal that occurs when a terminating exception is thrown during pipeline operations.
+    - If the dispose block has already been called (as it is after the `end{}` step), it will simply return immediately.
+
+### Behaviour
+
+The `dispose{}` block _does not_ guarantee output, but neither does it prevent it.
+
+- If output is received from the `dispose{}` block during normal command operation, it will be sent along the pipeline as per usual.
+- If output is received from the `dispose{}` block during a halting pipeline, it will be **discarded**.
```

Comment:
What use case is there for sending standard output along the pipeline from a `dispose` block? I would identify output from the `dispose` block as an anti-pattern, and since the `dispose` block is net new and not legacy, why not just avoid the anti-pattern in the first place and automatically suppress all _standard_ output from the `dispose` block. Other stream messages could be useful and should be allowed, as you have indicated below.

Reply:
I have yet to find a way to decouple standard output and error stream messages. If we suppress output completely, the error pipe is also not available, and as such the only error information that can be communicated is via terminating errors. This may not be desirable.

I agree it should be identified as an anti-pattern, but I don't think it's PS's purview to summarily prevent it. There may potentially be use cases for supplying some additional output at this step, perhaps _only_ supplying output during `dispose{]` to ensure that all operations completed successfully and with no errors before output can be received downstream.

Reply:
> I have yet to find a way to decouple standard output and error stream messages. If we suppress output completely, the error pipe is also not available, and as such the only error information that can be communicated is via terminating errors. This may not be desirable.

I agree non-terminating errors should be communicable. Someone just needs to figure out how to allow those messages to be processed without allowing standard output.

> I agree it should be identified as an anti-pattern, but I don't think it's PS's purview to summarily prevent it.

Sure it is. You're in `Dispose`. That block is specifically to facilitate releasing resources that are no longer needed. IMHO the community needs solid, unambiguous guidance on how to use `Dispose` rather than just another block where they can use it for its intent, or for other things if they want to muck around.

Reply:
Guidance, yes. Not sure it makes sense to impose hard limitations necessarily.

I'm not really against your position, more undecided here myself. 🤷‍♂

Reply:
The notion of output from `Dispose` is also broken IMHO, because `Dispose` only gets called when the _pipeline_ is complete (i.e. when the pipeline has finished processing, it calls `DoComplete` on each of the commands in the pipeline, which is where you would invoke `Dispose`, and at that point, if a command produces output from its `Dispose` block, where would that output go? Not to the next command, because it's already finished processing. So the notion of `Dispose` producing output that would go somewhere is really confusing.

I think having `Dispose` blocks only invoke when the pipeline is complete is absolutely the right place to invoke it, but given where it is invoked, it really shouldn't be allowed to produce standard output because that output would not be pipeline-able at that point, would it (I haven't tried your PR code, just read through some of it, so I haven't tested it to see)?

Reply:
Dispose gets invoked:

1. After _each command's_ `end{}` / `EndProcessing()`, _or_
2. For each command in turn when a terminating error occurs in the pipeline, as the pipeline is being disposed.

In the 1st case, output _could_ conceivably be passed. Not sure it's a great use case or something to be encouraged, but possible. 

In the 2nd, output cannot be passed, as the pipeline is already being disposed.

But yes, I agree that in general output does not work super well with the concept of `dispose{}`

Reply:
After `end`? That's different than how cmdlets work, right? Unless I'm mistaken, they dispose once the _pipeline_ has finished.

What if in my `begin` I open a connection to some remote resource, and I'm using a pipeline where commands I pipe to are going to use that connection? If I close that connection/dispose the resource in `dispose`, and if `dispose` is invoked after `end`, and let's just say I get creative and try to collect my output and send it down the pipeline in `dispose`, but after I close my connection. Odd scenario, I know, but outputting from `dispose` is odd no matter what, and I think there is a very good reason for cmdlets not disposing until the entire pipeline has completed. I expect functions to behave the same way. We really don't need yet-another-difference between advanced functions and cmdlets if you ask me.

Reply:
If the code is implemented as currently written, both cmdlets and functions will dispose after their `end{}` block rather than when the pipeline is disposed. 

Can you share an example of a command that exists using the semantics you describe? I'm not aware of any that would do this and also dispose of the connection themselves. It would generally be unfeasible to expect that the command _always_ be used in a pipeline context in such a manner, and would mean you cannot reliably use it outside of a pipeline context.

Reply:
I don't have an example...was just thinking that last bit out loud.

Regardless, I've said enough: I really don't like output from a `dispose` block, to the point where I would prevent it from being allowed if I could. I'll leave it at that.

Reply:
Appreciate the suggestions! 😊

Reply:
@KirkMunro FWIW I found a way to disable output from `dispose{}` so that change has been added. I'll double check and ensure it's mentioned in the text here as well, but I think I added that a while back.